src/core/thread/threadbase.d

```diff
@@ -0,0 +1,1374 @@
+/**
+ * The threadbase module provides OS-independent code
+ * for thread storage and management.
+ *
+ * Copyright: Copyright Sean Kelly 2005 - 2012.
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Sean Kelly, Walter Bright, Alex Rønne Petersen, Martin Nowak
+ * Source:    $(DRUNTIMESRC core/thread/osthread.d)
+ */
+
+module core.thread.threadbase;
+
+import core.thread.context;
+import core.thread.types;
+import core.time;
+import core.sync.mutex;
+import core.stdc.stdlib : free, realloc;
+
+private
+{
+    // interface to rt.tlsgc
+    import core.internal.traits : externDFunc;
+
+    alias rt_tlsgc_init = externDFunc!("rt.tlsgc.init", void* function() nothrow @nogc);
+    alias rt_tlsgc_destroy = externDFunc!("rt.tlsgc.destroy", void function(void*) nothrow @nogc);
+
+    alias ScanDg = void delegate(void* pstart, void* pend) nothrow;
+    alias rt_tlsgc_scan =
+        externDFunc!("rt.tlsgc.scan", void function(void*, scope ScanDg) nothrow);
+
+    alias rt_tlsgc_processGCMarks =
+        externDFunc!("rt.tlsgc.processGCMarks", void function(void*, scope IsMarkedDg) nothrow);
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Thread and Fiber Exceptions
+///////////////////////////////////////////////////////////////////////////////
+
+
+/**
+ * Base class for thread exceptions.
+ */
+class ThreadException : Exception
+{
+    @nogc @safe pure nothrow this(string msg, string file = __FILE__, size_t line = __LINE__, Throwable next = null)
+    {
+        super(msg, file, line, next);
+    }
+
+    @nogc @safe pure nothrow this(string msg, Throwable next, string file = __FILE__, size_t line = __LINE__)
+    {
+        super(msg, file, line, next);
+    }
+}
+
+
+/**
+* Base class for thread errors to be used for function inside GC when allocations are unavailable.
+*/
+class ThreadError : Error
+{
+    @nogc @safe pure nothrow this(string msg, string file = __FILE__, size_t line = __LINE__, Throwable next = null)
+    {
+        super(msg, file, line, next);
+    }
+
+    @nogc @safe pure nothrow this(string msg, Throwable next, string file = __FILE__, size_t line = __LINE__)
+    {
+        super(msg, file, line, next);
+    }
+}
+
+private
+{
+    // Handling unaligned mutexes are not supported on all platforms, so we must
+    // ensure that the address of all shared data are appropriately aligned.
+    import core.internal.traits : classInstanceAlignment;
+
+    enum mutexAlign = classInstanceAlignment!Mutex;
+    enum mutexClassInstanceSize = __traits(classInstanceSize, Mutex);
+
+    extern(C) void* swapContext(void* newContext) nothrow @nogc;
+
+    extern extern(C) immutable size_t threadSizeof;
+
+    extern(C) void* getStackBottom() nothrow @nogc;
+    extern(C) void* getStackTop() nothrow @nogc;
```

Comment:
`extern(C)` isn't a proper way to call functions from another module without importing it. These C functions are mostly used for functions invoked by the compiler; druntime-internal, there are helpers like `externDFunc`. `extern(D)` provides type safety and makes sure the functions don't clash with user code (that's why the C ones usually start with `_d_`).

IIRC, there are other places where similar stuff is handled by conceptually making `ThreadBase` a template taking a platform-specific `NativeContext` as template parameter, which exposes functionality like `getStack{Bottom,Top}()` etc. In the end, there'd be something like
```D
version (Windows)
    alias Thread = ThreadBase!WindowsContext;
else version (Posix)
    alias Thread = ThreadBase!PthreadContext;
...
```

Reply:
Ok, will substitute it by externDFunc.

About `NativeContext`: yes. But in this PR I tried to do everything gradually and make as few changes as possible, just copy and pasting.

Reply:
Done, except `scanWindowsOnly` and `swapContext`. For some reason Windows version isn't links if extern(D) used.
I investigate this problem.